File: container-di/src/test/scala/com/yahoo/container/di/ConfigRetrieverTest.scala
// Copyright 2017 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.container.di

import com.yahoo.config.test.{Bootstrap1Config, Bootstrap2Config, TestConfig}
import com.yahoo.container.di.ConfigRetriever.{BootstrapConfigs, ComponentsConfigs}
import com.yahoo.vespa.config.ConfigKey
import org.hamcrest.CoreMatchers.{is, instanceOf => hamcrestInstanceOf}
import org.hamcrest.Matcher
import org.junit.Assert._
import org.junit.{After, Before, Ignore, Test}

import scala.reflect.ClassTag
import scala.collection.JavaConverters._

/**
 *
 * @author gjoranv
 * @author tonytv
 */
class ConfigRetrieverTest {

  var dirConfigSource: DirConfigSource = null

  @Before def setup()  {
    dirConfigSource = new DirConfigSource("ConfigRetrieverTest-")
  }

  @After def cleanup() { dirConfigSource.cleanup() }

  @Test
  def require_that_bootstrap_configs_come_first() {
    writeConfigs()
    val retriever = createConfigRetriever()
    val bootstrapConfigs = retriever.getConfigs(Set(), 0)

    assertThat(bootstrapConfigs, instanceOf[BootstrapConfigs])
  }

  @Test
  def require_that_components_comes_after_bootstrap() {
    writeConfigs()
    val retriever = createConfigRetriever()
    val bootstrapConfigs = retriever.getConfigs(Set(), 0)

    val testConfigKey = new ConfigKey(classOf[TestConfig], dirConfigSource.configId)
    val componentsConfigs = retriever.getConfigs(Set(testConfigKey), 0)

    componentsConfigs match {
      case ComponentsConfigs(configs) => assertThat(configs.size, is(3))
      case _ => fail("ComponentsConfigs has unexpected type: " + componentsConfigs)
    }
  }

  @Test
  def require_no_reconfig_when_restart_on_redeploy() {
    // TODO
    writeConfigs()
    val retriever = createConfigRetriever()
    val bootstrapConfigs = retriever.getConfigs(Set(), 0)

    val testConfigKey = new ConfigKey(classOf[TestConfig], dirConfigSource.configId)
    val componentsConfigs = retriever.getConfigsOnce(Set(testConfigKey), 0, true)

    componentsConfigs match {
      case Some(snapshot) => fail("Expected no configs")
      case _ => // ok
    }
  }

  @Test(expected = classOf[IllegalArgumentException])
  @Ignore
  def require_exception_upon_modified_components_keys_without_bootstrap() {
    writeConfigs()
    val retriever = createConfigRetriever()
    val testConfigKey = new ConfigKey(classOf[TestConfig], dirConfigSource.configId)
    val bootstrapConfigs = retriever.getConfigs(Set(), 0)
    val componentsConfigs = retriever.getConfigs(Set(testConfigKey), 0)
    retriever.getConfigs(Set(testConfigKey, new ConfigKey(classOf[TestConfig],"")), 0)
  }

  @Test
  def require_that_empty_components_keys_after_bootstrap_returns_components_configs() {
    writeConfigs()
    val retriever = createConfigRetriever()
    assertThat(retriever.getConfigs(Set(), 0), instanceOf[BootstrapConfigs])
    assertThat(retriever.getConfigs(Set(), 0), instanceOf[ComponentsConfigs])
  }

  def writeConfigs() {
    writeConfig("bootstrap1", """dummy "ignored" """")
    writeConfig("bootstrap2", """dummy "ignored" """")
    writeConfig("test", """stringVal "ignored" """")
  }

  def createConfigRetriever() = {
    val configId = dirConfigSource.configId
    val subscriber = new CloudSubscriberFactory(dirConfigSource.configSource)
    new ConfigRetriever(
      Set(new ConfigKey(classOf[Bootstrap1Config], configId),
          new ConfigKey(classOf[Bootstrap2Config], configId)),
      (keys) => subscriber.getSubscriber(keys.asJava))
  }

  def writeConfig = dirConfigSource.writeConfig _

  def instanceOf[T: ClassTag] = hamcrestInstanceOf(implicitly[ClassTag[T]].runtimeClass): Matcher[AnyRef]
}

File: container-di/src/test/java/com/yahoo/container/di/ConfigRetrieverTest.java
// Copyright 2018 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.container.di;

import com.yahoo.config.ConfigInstance;
import com.yahoo.config.test.Bootstrap1Config;
import com.yahoo.config.test.Bootstrap2Config;
import com.yahoo.config.test.TestConfig;
import com.yahoo.container.di.ConfigRetriever.BootstrapConfigs;
import com.yahoo.container.di.ConfigRetriever.ComponentsConfigs;
import com.yahoo.container.di.ConfigRetriever.ConfigSnapshot;
import com.yahoo.vespa.config.ConfigKey;
import org.hamcrest.Matchers;
import org.junit.After;
import org.junit.Before;
import org.junit.Ignore;
import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.ExpectedException;

import java.util.Collections;
import java.util.HashSet;
import java.util.Optional;
import java.util.Set;

import static org.hamcrest.CoreMatchers.instanceOf;
import static org.hamcrest.CoreMatchers.is;
import static org.junit.Assert.assertThat;
import static org.junit.Assert.fail;

/**
 *
 * @author gjoranv
 * @author Tony Vaagenes
 * @author ollivir
 */
public class ConfigRetrieverTest {

    private DirConfigSource dirConfigSource = null;

    @Before
    public void setup() {
        dirConfigSource = new DirConfigSource("ConfigRetrieverTest-");
    }

    @After
    public void cleanup() {
        dirConfigSource.cleanup();
    }

    @Test
    public void require_that_bootstrap_configs_come_first() {
        writeConfigs();
        ConfigRetriever retriever = createConfigRetriever();
        ConfigSnapshot bootstrapConfigs = retriever.getConfigs(Collections.emptySet(), 0, false);

        assertThat(bootstrapConfigs, Matchers.instanceOf(BootstrapConfigs.class));
    }

    @Test
    @SuppressWarnings("unused")
    public void require_that_components_comes_after_bootstrap() {
        writeConfigs();
        ConfigRetriever retriever = createConfigRetriever();
        ConfigSnapshot bootstrapConfigs = retriever.getConfigs(Collections.emptySet(), 0, false);

        ConfigKey<? extends ConfigInstance> testConfigKey = new ConfigKey<>(TestConfig.class, dirConfigSource.configId());
        ConfigSnapshot componentsConfigs = retriever.getConfigs(Collections.singleton(testConfigKey), 0);

        if (componentsConfigs instanceof ComponentsConfigs) {
            assertThat(componentsConfigs.size(), is(3));
        } else {
            fail("ComponentsConfigs has unexpected type: " + componentsConfigs);
        }
    }

    @Test
    @SuppressWarnings("unused")
    public void require_no_reconfig_when_restart_on_redeploy() {
        // TODO
        writeConfigs();
        ConfigRetriever retriever = createConfigRetriever();
        ConfigSnapshot bootstrapConfigs = retriever.getConfigs(Collections.emptySet(), 0, false);

        ConfigKey<? extends ConfigInstance> testConfigKey = new ConfigKey<>(TestConfig.class, dirConfigSource.configId());
        Optional<ConfigSnapshot> componentsConfigs = retriever.getConfigsOnce(Collections.singleton(testConfigKey), 0, true);

        if (componentsConfigs.isPresent()) {
            fail("Expected no configs");
        }
    }

    @Rule
    public ExpectedException expectedException = ExpectedException.none();

    @Ignore
    @SuppressWarnings("unused")
    public void require_exception_upon_modified_components_keys_without_bootstrap() {
        expectedException.expect(IllegalArgumentException.class);

        writeConfigs();
        ConfigRetriever retriever = createConfigRetriever();
        ConfigKey<? extends ConfigInstance> testConfigKey = new ConfigKey<>(TestConfig.class, dirConfigSource.configId());
        ConfigSnapshot bootstrapConfigs = retriever.getConfigs(Collections.emptySet(), 0, false);
        ConfigSnapshot componentsConfigs = retriever.getConfigs(Collections.singleton(testConfigKey), 0, false);
        Set<ConfigKey<? extends ConfigInstance>> keys = new HashSet<>();
        keys.add(testConfigKey);
        keys.add(new ConfigKey<>(TestConfig.class, ""));
        retriever.getConfigs(keys, 0);
    }

    @Test
    public void require_that_empty_components_keys_after_bootstrap_returns_components_configs() {
        writeConfigs();
        ConfigRetriever retriever = createConfigRetriever();
        assertThat(retriever.getConfigs(Collections.emptySet(), 0), instanceOf(BootstrapConfigs.class));
        assertThat(retriever.getConfigs(Collections.emptySet(), 0), instanceOf(ComponentsConfigs.class));
    }

    public void writeConfigs() {
        writeConfig("bootstrap1", "dummy \"ignored\"");
        writeConfig("bootstrap2", "dummy \"ignored\"");
        writeConfig("test", "stringVal \"ignored\"");
    }

    private ConfigRetriever createConfigRetriever() {
        String configId = dirConfigSource.configId();
        CloudSubscriberFactory subscriber = new CloudSubscriberFactory(dirConfigSource.configSource());
        Set<ConfigKey<? extends ConfigInstance>> keys = new HashSet<>();
        keys.add(new ConfigKey<>(Bootstrap1Config.class, configId));
        keys.add(new ConfigKey<>(Bootstrap2Config.class, configId));
        return new ConfigRetriever(keys, keySet -> subscriber.getSubscriber(keySet));
    }

    private void writeConfig(String name, String contents) {
        dirConfigSource.writeConfig(name, contents);
    }
}
